### PR TITLE
BUG: Infer return types for Fortran functions in `f2py`

### LINKED ISSUE #23598: BUG: f2py does not properly wrap functions
### Describe the issue:

I confess I can't determine if this issue is a bug or an implicit lack of feature (partial inability to deal with functions), which does _not_ raise an error, yet produces strange results. 

f2py works as as expected for subroutines. In contrast, with functions there are 2 issues:

* When functions are defined using the syntax `result-type function function-name(arguments, ) result (result-name)`, the result-type is always interpreted as real. There is a type mismatch warning. The function "works", but "just" returns the wrong type.
 
* When functions are defined using the classical syntax `function function-name(arguments, )`, the result-type is correctly interpreted, but f2py raises a warning about the number of arguments. Despite the warning, the function seems to work well.

An example can be found [here](https://github.com/HugoMVale/fortran-in-python/tree/main/src/example-f2py).

### Reproduce the code example:

```python
# original fortran functions

integer function intproduct(a, b) result(res)
  !! Product of two integers
  !! This will not work as as expected
  integer, intent(in) :: a, b
  res = a*b
end function

function intproduct2(a, b)
  !! Product of two integers
  !! This will not work as as expected
  integer, intent(in) :: a, b
  integer :: intproduct2
  intproduct2 = a*b
end function
```

### Error message:

```shell
# This is not the error message (there is no error), but the intermediate pyf file

function intproduct(a,b) result (res) ! in :fortranmodule:mathtools.f90:mathtools
  integer intent(in) :: a
  integer intent(in) :: b
  real :: res                        ! wrong
end function intproduct
function intproduct2(a,b) ! in :fortranmodule:mathtools.f90:mathtools
  integer intent(in) :: a
  integer intent(in) :: b
  integer :: intproduct2             ! ok
end function intproduct2
```

### Runtime information:

```
C:\Users\X\AppData\Local\Temp\tmpdov3t9ts\src.win-amd64-3.11\fortranmodule-f2pywrappers2.f90:5:46:

    5 |       subroutine f2pywrap_mathtools_intproduct (intproductf2pywrap, a, b&
      |                                              1
......
   44 |       subroutine f2pywrap_mathtools_intproduct (intproductf2pywrap, intp&
      |                                              2
Warning: Type mismatch in argument 'intproduct' (REAL(4)/INTEGER(4)) between (1) and (2)
C:\Users\X\AppData\Local\Temp\tmpdov3t9ts\src.win-amd64-3.11\fortranmodule-f2pywrappers2.f90:13:47:

   13 |       subroutine f2pywrap_mathtools_intproduct2 (intproduct2f2pywrap, a,&
      |                                               1
......
   51 |       subroutine f2pywrap_mathtools_intproduct2 (intproduct2f2pywrap, in&
      |                                               2
Warning: 'f2pywrap_mathtools_intproduct2' has the wrong number of arguments between (1) and (2)
```

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by HaoZeke**
Closes #23598.

### CODE REVIEW

**Comment by seberg** [comment]
This looks good, but I don't really know fortran.  @melissawm could you have a brief look?

Should this be backported?

File: doc/source/f2py/code/ftype.f [hunk 1]
```diff
@@ -4,6 +4,6 @@ SUBROUTINE FOO(N)
 Cf2py integer optional,intent(in) :: n = 13
       REAL A,X
       COMMON /DATA/ A,X(3)
-      PRINT*, "IN FOO: N=",N," A=",A," X=[",X(1),X(2),X(3),"]"
+C     PRINT*, "IN FOO: N=",N," A=",A," X=[",X(1),X(2),X(3),"]"
```
Comment (melissawm) [inline doc/source/f2py/code/ftype.f:7]: Maybe we want to remove this line entirely?

**Review by melissawm — APPROVED** [review]
This looks good to me, @HaoZeke - thanks! I also checked against SciPy and it looks good.

**Comment by seberg** [comment]
Thanks, will just put it in tomorrow (or someone else please do)!  I don't see an obvious reason for backport anywhere, but if there is, please label it.

**Comment by seberg** [comment]
Thanks again @HaoZeke.

File: doc/source/f2py/code/ftype.f (hunk 1 above)
Comment (HaoZeke) [reply doc/source/f2py/code/ftype.f:7]: Ack, sorry, didn't see this earlier, I'll open a general cleanup to make sure we have no print statements in the `f2py` tests this weekend.

**Pull request merged**